src/algebra/hom/centroid.lean

```diff
@@ -0,0 +1,299 @@
+
+instance : has_mul (centroid_hom α) := ⟨comp⟩
+
+instance has_smul_nat : has_smul ℕ (centroid_hom α) :=
+⟨λ n f,
+  { map_mul_left' := λ a b,
+      by { change n • f (a * b) = a * n • f b, rw [map_mul_left f, ←mul_smul_comm] },
+    map_mul_right' := λ a b,
+      by { change n • f (a * b) = n • f a * b, rw [map_mul_right f, ←smul_mul_assoc] },
+    .. (n • f : α →+ α) }⟩
+
+instance has_npow_nat : has_pow (centroid_hom α) ℕ :=
+⟨λ f n,
+{  map_mul_left' := λ a b, begin
+    induction n with n ih,
+    { simp },
+    { rw pow_succ,
+      exact (congr_arg f.to_End ih).trans (f.map_mul_left' _ _) }
+  end,
+  map_mul_right' := λ a b, begin
+    induction n with n ih,
+    { simp },
+    { rw pow_succ,
+      exact (congr_arg f.to_End ih).trans (f.map_mul_right' _ _) }
+  end,
+  ..(f.to_End ^ n : add_monoid.End α) }⟩
+
+@[simp] lemma coe_zero : ⇑(0 : centroid_hom α) = 0 := rfl
+@[simp] lemma coe_one : ⇑(1 : centroid_hom α) = id := rfl
+@[simp] lemma coe_add (f g : centroid_hom α) : ⇑(f + g) = f + g := rfl
+@[simp] lemma coe_mul (f g : centroid_hom α) : ⇑(f * g) = f ∘ g := rfl
+-- Eligible for `dsimp`
+@[simp, nolint simp_nf] lemma coe_nsmul (f : centroid_hom α) (n : ℕ) : ⇑(n • f) = n • f := rfl
+
+@[simp] lemma zero_apply (a : α) : (0 : centroid_hom α) a = 0 := rfl
+@[simp] lemma one_apply (a : α) : (1 : centroid_hom α) a = a := rfl
+@[simp] lemma add_apply (f g : centroid_hom α) (a : α) : (f + g) a = f a + g a := rfl
+@[simp] lemma mul_apply (f g : centroid_hom α) (a : α) : (f * g) a = f (g a) := rfl
+-- Eligible for `dsimp`
+@[simp, nolint simp_nf]
+lemma nsmul_apply (f : centroid_hom α) (n : ℕ) (a : α) : (n • f) a = n • f a := rfl
+
+-- cf.`add_monoid_hom.add_comm_monoid`
+instance : add_comm_monoid (centroid_hom α) :=
+coe_to_add_monoid_hom_injective.add_comm_monoid _ rfl (λ _ _, rfl) (λ _ _, rfl)
+
+instance : has_nat_cast (centroid_hom α) :=
+{ nat_cast := λ n, n • 1 }
+
+@[simp] lemma nat_cast_apply (n : ℕ) (m : α):
+  (n : centroid_hom α) m = n • m := rfl
+
+-- cf `add_monoid.End.semiring`
+instance : semiring (centroid_hom α) :=
+to_End_injective.semiring _ (by { ext, refl }) (by { ext, refl }) (λ _ _, rfl) (λ _ _, rfl)
+  (λ _ _, rfl) (λ _ _, by { ext, refl }) (λ n, rfl)
+
+lemma comp_mul_comm (T S : centroid_hom α) (a b : α) : (T ∘ S) (a * b) = (S ∘ T) (a * b) :=
+by rw [comp_app, map_mul_right, map_mul_left, ←map_mul_right, ←map_mul_left]
+
+end non_unital_non_assoc_semiring
+
+section non_unital_non_assoc_ring
+variables [non_unital_non_assoc_ring α]
+
+/-- Negation of `centroid_hom`s as a `centroid_hom`. -/
+instance : has_neg (centroid_hom α) :=
+⟨λ f,
+  { map_mul_left' := by simp [map_mul_left],
+    map_mul_right' := by simp [map_mul_right],
+    .. (-f : α →+ α) }⟩
+
+
+instance : has_sub (centroid_hom α) :=
+⟨λ f g,
+{ map_mul_left' := λ a b, by simp [map_mul_left, mul_sub],
+  map_mul_right' := λ a b, by simp [map_mul_right, sub_mul],
+  .. (f - g : α →+ α) }⟩
+
+instance has_smul_int : has_smul ℤ (centroid_hom α) :=
```

Comment:
```suggestion
instance has_zsmul : has_smul ℤ (centroid_hom α) :=
```

Reply:
If I understood @eric-wieser correctly, he wanted the names to match the typeclass name https://github.com/leanprover-community/mathlib/pull/12746#pullrequestreview-1060159820 ?

Reply:
I think we have more instances that use the long names (the PR) instead of the short name (the suggestion), so would be inclined to leave things as is.